### PR TITLE
Fix ``used-before-assignment`` for assignment expressions in lambda

### LINKED ISSUE #3877: False positive undefined-variable when a function attribute of same name is set
### Steps to reproduce
```python
CONDITION = None
FUNC = lambda: CONDITION
FUNC.CONDITION = 42
```
Pylint output:
```
pylint_bug_repro.py:2:15: E0602: Undefined variable 'CONDITION' (undefined-variable)
```

### Current behavior
An "undefined variable" is incorrectly reported.

### Expected behavior
Zero issue detected when defining a [function attribute](https://www.python.org/dev/peps/pep-0232/)
with the same name of a local variable in the current scope.

### pylint --version output
```
$ pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.7.7 (default, Apr 10 2020, 07:59:19)
[GCC 9.3.0]
```

### Additional notes
There are a few other existing issues related to "false positive undefined-variable" errors:
- https://github.com/PyCQA/pylint/issues/2491
- https://github.com/PyCQA/pylint/issues/3688
- https://github.com/PyCQA/pylint/issues/3791

### LINKED ISSUE #5360: False positive: E0602 undefined variable with assignment expression
### Bug description

The following snippet produces `E0602: Undefined variable 'x_0' (undefined-variable)` as output from pylint.

```py
sorted_things = sorted(
        things,
        key=lambda thing: x_0 
        if (x_0 := thing.this_value)
        < (x_1 := thing.that_value)
        else x_1,
)
```

### Configuration

_No response_

### Command used

```shell
Basic `pylint` call as a pre-commit hook.
```

### Pylint output

```shell
pylint...................................................................Failed
- hook id: pylint
- exit code: 2

************* Module utils
project_dir/utils.py:197:22: E0602: Undefined variable 'x_0' (undefined-variable)
```

### Expected behavior

Should be no error.

### Pylint version

```shell
pylint --version
pylint 2.11.1
astroid 2.8.4
Python 3.9.5 (default, May  4 2021, 03:36:27) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

MacOS 11.5.2

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

This closes #5360, closes #3877

See a prior discussion in https://github.com/PyCQA/pylint/pull/5241#issuecomment-980041659 and @jacobtylerwalls's comments about it https://github.com/PyCQA/pylint/pull/5241#issuecomment-980806428.

This also fixes the message that gets raised in `home-assistant` https://github.com/home-assistant/core/blob/296b73874054eec31d91f0db1b9a625ad5d47c32/homeassistant/components/sht31/sensor.py#L63. That also seems to indicate to me that this code is not doing anything meaningful.

The original commit that added it was https://github.com/PyCQA/pylint/commit/5193b7d7fd63b5ddc4077cb5f37c3cd55a9d6561.
The tests added in that commit still pass with the change in this PR so I think this should be fine?

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]
There's a pattern of simply removing code and fixing bug, hmm. 🤔

**Comment by DanielNoord** [comment]
Also closes #3877.

Adding it to the list of closed issues.

File: ChangeLog
Comment (DanielNoord) [inline ChangeLog:40]:
```suggestion
  Closes #5360, #3877
```

File: doc/whatsnew/2.13.rst
Comment (DanielNoord) [inline doc/whatsnew/2.13.rst:75]:
```suggestion
  Closes #5360, #3877
```

**Comment by Pierre-Sassoulas** [comment]
Do you want another review before merging ?

**Comment by DanielNoord** [comment]
I think this should be fine. We have two separate issues reports about this and this fixes this issue. I don't think this was an intended message.

**Pull request merged**
